find_entry: Entry lookup support

tarcieri opened the pull request:
Not completely finished, but running into a confusing borrow checker error

tarcieri committed:
find_entry: Entry lookup support

Extracts an inode-like Node type, and supports traversing paths by
looking up child nodes while iterating over the path.

carllerche commented:
You can get it to compile by changing the following line in `lib.rs:191`

```diff
-    pub fn find_entry<'a>(&'a self, txn: &'a mut RwTransaction<'a>, path: &str) -> Result<Entry> {
+    pub fn find_entry<'a>(&'a self, txn: &'a mut RwTransaction, path: &str) -> Result<Entry> {
```

When you use explicit lifetimes, you are saying that they must be *exactly* the same which is what was causing the problem. By not specifying them, you are letting rustc infer.

tarcieri committed:
Extract Id type

Creates a tuple struct for wrapping Ids with several convenience functions.

tarcieri committed:
DRY out Node and Entry

Entry contained many of the same fields as Node. This stores a Node
in each Entry so we don't need to duplicate these fields.
